### DISCUSSION TITLE
TRMM Agent on QNAP NAS

### DISCUSSION THREAD

**Discussion by janstalhut** [amidaware/tacticalrmm#1971] [Q&A]
The Linux agent already works very well on Snology NAS systems. But we also have some customers who use NAS systems from the manufacturer QNAP. If I try to install the Linux agent (ARM or x86_64, depending on the model), the script immediately runs into an error.

I get the following output on a system with an x86_64 processor:
```
ps: unrecognized option '--no-headers'
BusyBox v1.24.1 (2024-07-12 03:12:23 CST) multi-call binary.

Usage: ps [-o COL1,COL2=HEADER] [-T]

Show list of processes

	-o COL1,COL2=HEADER	Select columns for display
	-T			Show threads

QNAP Options:
	--columns N

This install script only supports systemd
Please install systemd or manually create the service using your systems's service manager
```

On a system with an ARM 32 bit (arm7l), however, I only get the following output:
```
This install script only supports systemd
Please install systemd or manually create the service using your systems's service manager
```

Is there any way to install the agent on a QNAP NAS anyway? Has anyone done that?

What would be the concrete way behind the note "manually create the service using your system's service manager"?

Does anyone have any ideas what else I could try? If you need any more technical details about the device, I'm happy to provide them.

**Comment by dinger1986** [comment]
I dont know if anyone has done it but essentially you need to edit the install script to remove the requirement for systemd, allow it to install and create a new file `/etc/init.d/tacticalagent`.

It should contain the following:

```
#!/bin/sh
### BEGIN INIT INFO
# Provides:
# Default-Start:     3
# Default-Stop:      0 1 6
# Short-Description: Start daemon at boot time
# Description:       Enable service provided by daemon.
### END INIT INFO

export PATH=$PATH:/usr/local/bin

cmd="tacticalagent -m svc &"
user="root"

name="tacticalagent"
pid_file="/var/run/$name.pid"
stdout_log="/var/log/$name.log"
stderr_log="/var/log/$name.err"

get_pid() {
    cat "$pid_file"
}

is_running() {
    [ -f "$pid_file" ] && ps -p `get_pid` > /dev/null 2>&1
}

case "$1" in
    start)
    if is_running; then
        echo "Already started"
    else
        echo "Starting $name"
        cd "$dir"
        if [ -z "$user" ]; then
            $cmd >> "$stdout_log" 2>> "$stderr_log" &
        else
			$cmd >> "$stdout_log" 2>> "$stderr_log" &
        fi
        echo $! > "$pid_file"
        if ! is_running; then
            echo "Unable to start, see $stdout_log and $stderr_log"
            exit 1
        fi
    fi
    ;;
    stop)
    if is_running; then
        echo -n "Stopping $name.."
        kill `get_pid`
        for i in 1 2 3 4 5 6 7 8 9 10
        # for i in `seq 10`
        do
            if ! is_running; then
                break
            fi

            echo -n "."
            sleep 1
        done
        echo

        if is_running; then
            echo "Not stopped; may still be shutting down or shutdown may have failed"
            exit 1
        else
            echo "Stopped"
            if [ -f "$pid_file" ]; then
                rm "$pid_file"
            fi
        fi
    else
        echo "Not running"
    fi
    ;;
    restart)
    $0 stop
    if is_running; then
        echo "Unable to stop, will not attempt to start"
        exit 1
    fi
    $0 start
    ;;
    status)
    if is_running; then
        echo "Running"
    else
        echo "Stopped"
        exit 1
    fi
    ;;
    *)
    echo "Usage: $0 {start|stop|restart|status}"
    exit 1
    ;;
esac

exit 0
```

You might then need to install or recover the meshagent to get remote background working.

You can then start it with `/etc/init.d/tacticalagent start`, the init.d script doesnt work 100% as doesnt report if the agents running etc but it works well enough for having the agent running so well enough for the few machines I have that are still running init.d

**Reply by janstalhut** [reply]
Thanks for the script. It basically worked on QNAP NAS of various generations. Unfortunately, the basic problem here is that all changes in /etc, /usr, /opt etc. do not survive a system restart. The entire system is apparently booted in a read-only mode in a RAM disk.

The very first solution would probably be to create a special package that is installed as an app in the QNAP system and as such can then officially store an autostart script and/or services.

**Reply by dinger1986** [reply]
have you looked at https://forum.qnap.com/viewtopic.php?t=129247
